deepspeed/runtime/zero/stage3.py

```diff
@@ -2832,6 +2832,17 @@ def save_checkpoint_prologue(self):
     def save_checkpoint_epilogue(self):
         self.persistent_parameters[0].all_gather(self.persistent_parameters)
 
+    def save_partitioned_weights(self, state_dict):
+        for name, param in self.module.named_parameters():
+            if name in state_dict.keys():
+                state_dict[name] = param.ds_tensor

```

Comment:
Found an issue here: `param.ds_tensor` is a flattened buffer. So `state_dicts` ends up having 1D vectors.

Reply:
but we can't shape it back to the original since we only have a part of the tensor, so doing something like `narrow(0, 0, param.ds_numel).view(param.ds_shape)` from `_allgather_param()` won't work and the shape has no meaning here anyway.

So this line of logic is useful when it's used to load the `param.ds_tensor` directly by each gpu, as coded in the rest of this PR.

I just tried to use it to get the partitioned fp16 weights, but now I understand this is not possible using this approach.

Bottom line - there is no problem here, just needed to understand that this is not a real `state_dict` that is being saved but something like `flattened_params_state_dict`.

All is good!